library/Scripting/Duktape/Raw.hs

```diff
@@ -144,12 +176,34 @@ foreign import capi safe "duktape.h duk_push_context_dump"
 
 -------------------------------------------------------------------------------------------------------
 
-createHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → Ptr () → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
+createHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → InternalUData → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
 createHeap allocf reallocf freef udata fatalf = do
-  ptr ← c_duk_create_heap allocf reallocf freef udata fatalf
+  ptr ← c_duk_create_heap allocf reallocf freef (getInternalUData udata) fatalf
   if ptr /= nullPtr
      then newForeignPtr ptr (c_duk_destroy_heap ptr) >>= newMVar >>= return . Just
      else return Nothing
 
 createHeapF ∷ FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
-createHeapF = createHeap nullFunPtr nullFunPtr nullFunPtr nullPtr
+createHeapF = createHeap nullFunPtr nullFunPtr nullFunPtr (InternalUData nullPtr)
+
+-- | A TimeoutCheck is an IO action that returns True when the current script evaluation
+-- should timeout (interpreter throws RangeError).
+createGovernedHeap ∷ FunPtr DukAllocFunction → FunPtr DukReallocFunction → FunPtr DukFreeFunction → TimeoutCheck → FunPtr DukFatalFunction → IO (Maybe DuktapeCtx)
+createGovernedHeap allocf reallocf freef timeoutCheck fatalf = do
+  (udata, release) ← wrapTimeoutCheckUData timeoutCheck
+  mctx ← createHeap allocf reallocf freef udata fatalf
```

Comment:
If the thread is killed after line 193, then what you just allocated won't ever be released.

Reply:
I think this is an issue in `createHeap` as well for the actual Duktape heap.